design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+2. Multiple credentials sets for virtual chunk resolution, accessing some chunks may require a different set than others.
+3. A mechanism to know all the "dependencies" of a repo. That is, all the "places" where its virtual chunks are hosted. This is important for users to know if they can delete old data, or if there are virtual datasets that depend on it.
+4. If the file hosting a virtual chunk changed after the chunk ref was written, we want to fail with a runtime error instead of serving corrupted data. This is because offset and length are usually not maintained during updates.
+
+## Discussion
+
+### Current virtual chunk URL styles
+
+* The open source stack seating on top of Icechunk (VirtualiZarr, etc.) uses URLs to identify virtual chunk locations.
+* One common case is many virtual chunks from a few objects.
+  * Example: `s3://some-bucket/some-prefix/some-file.nc` is pointed to by 10k chunks.
+  * Having between 10 and 10k chunks from the same object is not unusual
+* Other common use case for virtual zarr datasets: each chunk comes from a different object. This can happen, for example, with a virtual zarr dataset.
+  * Example: `s3://some-bucket/some-prefix/c/0/0/1` is pointed to by only 1 chunk. But there are many objects like it.
+  * There could be tens of millions of chunks (so tens of millions of different urls)
+
+### Supporting multiple virtual chunk locations
+
+Because of requirement 1, the same repo can contain virtual chunks in multiple buckets, object stores, and platforms, potentially even buckets with the same name in different platforms.
+
+When the user adds a virtual ref, they will need to indicate the details of the platform for Icechunk to be able to retrieve it. These includes what cloud it is, region, endpoint url, etc. This means, the current approach of pointing to a chunk with a simple URL is not going to be enough. Not without encoding several new attributes in the URL, which seems hacky. A URL such as `s3://foo/bar.nc`, in itself doesn't have enough details to reach the object. Even if the protocol part of that URL is identified with the wire protocol needed to talk to the object store, there are missing details such as the region, the endpoint, or the authentication mechanism.
+
+This all means, the following function may have to change:
+
+```python
+def set_virtual_ref(
+    self, key: str, location: str, offset: int, length: int
+) -> None: ...
+```
+
+### Credentials
+
+Credentials are a read time concern. When the user writes the virtual chunk, they don't necessarily know what credentials set will be used to read it.
+
+Because of requirements 1 and 2, when users open a repo they need to know what sets of credentials they have to provide to be able to access all the chunks. This is not immediately obvious, Icechunk should provide a way to introspect a repo and inform the user what credentials they need. Then the user can setup the Repo instance appropriately with the credentials.
```

Comment:
```suggestion
Because of requirements 1 and 2, when users open a repo they need to know what sets of credentials they have to provide to be able to access all the chunks. Icechunk should provide a way to introspect a repo and inform the user what credentials they need. Then the user can set up the Repo instance appropriately with the credentials.

Icechunk itself should never store credentials in any of its metadata files.
```